File: lib/stack/stack.js
(function () {
  'use strict';

  var Stack = function () {
    this.head = null;
  };

  Stack.prototype.isEmpty = function () {
    return this.head === null;
  };

  Stack.prototype.push = function (value) {
    this.head = { value: value, next: this.head };

    return this;
  };

  Stack.prototype.pop = function () {
    var stackValue;

    if (this.isEmpty()) return null;

    stackValue = this.head.value;
    this.head = this.head.next;

    return stackValue;
  };

  module.exports = Stack;
})();

File: lib/stack/stack.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "stack.h"

typedef struct Node {
	struct Node *next;
	int value;
} Node;

typedef struct Stack {
	Node *head;
} Stack;

Stack *stack_create()
{
	Stack *stack = malloc(sizeof(Stack));

	stack->head = NULL;

	return stack;
}

void stack_destroy(Stack *stack)
{
	Node *temp;

	while (!stack_is_empty(stack)) {
		temp = stack->head->next;
		free(stack->head);
		stack->head = temp;
	}

	free(stack);
}

bool stack_is_empty(Stack *stack)
{
	return stack->head == NULL;
}

Stack *stack_push(Stack *stack, int value)
{
	Node *head = malloc(sizeof(Node));

	head->value = value;
	head->next = stack->head;
	stack->head = head;

	return stack;
}

int stack_pop(Stack *stack)
{
	if (stack_is_empty(stack)) return 0;

	int value = stack->head->value;
	Node *temp = stack->head;
	stack->head = temp->next;
	free(temp);

	return value;
}
